File: 01/rust01/src/bin/part2.rs
fn main() {
    let input = include_str!("../../../input/part1.txt");

    let answer = process(input);

    println!("Answer: {}", answer);
}

fn process(input: &str) -> u32 {
    let nums_as_words = [
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];
    let answer = input
        .lines()
        .map(|l| {
            let mut nums = Vec::new();

            for (i, c) in l.chars().enumerate() {
                if let Some(num) = c.to_digit(10) {
                    nums.push(num);
                    continue;
                }

                let pos = nums_as_words.iter().position(|a| l[i..].starts_with(a));
                if let Some(pos) = pos {
                    nums.push((pos + 1).try_into().unwrap());
                }
            }

            if !nums.is_empty() {
                nums[0] * 10 + nums[nums.len() - 1]
            } else {
                println!("We got a 0");
                0
            }
        })
        .sum();

    answer
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn part_2_is_correct() {
        let input = "two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen";

        assert_eq!(process(input), 281)
    }
}

File: 01/go01/part2/part2.go
package part2

import (
	"strconv"
	"strings"
)

func Process(input string) int {
	intsAsNums := []string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}

	lines := strings.Split(input, "\n")
	answer := 0
	for _, line := range lines {
		chars := strings.Split(line, "")
		nums := []int{}
		for i, c := range chars {
			for j, ias := range intsAsNums {
				if strings.HasPrefix(line[i:], ias) {
					nums = append(nums, j+1)
				}
			}

			k, err := strconv.Atoi(c)
			if err != nil {
				continue
			}
			nums = append(nums, k)
		}

		if len(nums) > 0 {
			answer += nums[0]*10 + nums[len(nums)-1]
		}
	}

	return answer
}
